### PR TITLE
Oauth

### LINKED ISSUE #44: OAUTH login for GitHub/GitLab/Codeberg
As I'm working on this. Just having a issue :)

Simple flow: Ensure security and possible support PKCE extension for supported services. 

Do some front end to input nickname + password and than let continue the basic register flow by email verification(or not?)

### PR DESCRIPTION
**Pull request by Gusted**
Resolves #44

### CODE REVIEW

File: utils/oauth.go [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+	"net/url"
+	"strings"
+
+	"userstyles.world/config"
+)
+
+var (
+	githubClientID   = "f3f2d378d88794062895"
+	gitlabClientID   = "f906330603e645d54184fd13337ca51b762ea5da2188e7f248e109c686940897"
+	codebergClientID = "308920d9-e49d-4db6-baa9-7513c9f6dea2"
```
Comment (vednoc) [inline utils/oauth.go:16]: Should this also be turned into environment variables?

File: utils/oauth.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+		body = GiteaLikeAccessJson{
+			ClientID:     codebergClientID,
+			ClientSecret: config.CODEBERG_CLIENT_SECRET,
+			Code:         tempCode,
+			GrantType:    "authorization_code",
+			RedirectURI:  "http://localhost:3000/api/callback/codeberg/",
+		}
+	} else {
+		return OAuthResponse{}
+	}
```
Comment (vednoc) [inline utils/oauth.go:129]: Use a switch statement?

File: utils/oauth.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+	var userEndpoint string
+	if service == "github" {
+		userEndpoint = "https://api.github.com/user"
+	} else if service == "gitlab" {
+		userEndpoint = "https://gitlab.com/api/v4/user"
+	} else if service == "codeberg" {
+		userEndpoint = "https://codeberg.org/api/v1/user"
+	}
```
Comment (vednoc) [inline utils/oauth.go:171]: Use a switch statement?

File: views/register.html [hunk 4]
```diff
@@ -15,6 +15,26 @@ <h1>Register</h1>
                 {{ template "partials/alert" . }}
 
                 <section class="ta:c">
+                    <a href="http://localhost:3000/oauth/github" class="button-link">
+                        <button class="github oauth">
+                            <svg height="32" width="32" viewBox="0 0 16 16">
+                                <path fill-rule="evenodd" d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0016 8c0-4.42-3.58-8-8-8z"></path>
+                            </svg>
```
Comment (vednoc) [inline views/register.html:22]: Use a partial template.

File: utils/oauth.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+	} else if service == "codeberg" {
+		// Base URL.
+		oauthURL = "https://codeberg.org/login/oauth/authorize"
+		// Add our app client ID.
+		oauthURL += "?client_id=" + codebergClientID
+		// Define we want a code back
+		oauthURL += "&response_type=code"
+	} else {
+		return ""
+	}
```
Comment (vednoc) [inline utils/oauth.go:73]: Use a switch statement?

**Review by vednoc — COMMENTED** [review]
I'll test functionality tomorrow. This is what caught my attention for now.

File: utils/oauth.go (hunk 1 above)
Comment (Gusted) [reply utils/oauth.go:16]: It can, but they aren't a secret and just hardcoded 🤷🏽‍♀️

Comment (vednoc) [reply utils/oauth.go:16]: I was thinking of using different client IDs in development and production environments.

**Comment by Gusted** [comment]
Done +1

**Review by vednoc — APPROVED** [review]

**Review by vednoc — APPROVED** [review]

**Pull request merged**